pallet/staking/src/lib.rs

```diff
@@ -260,6 +340,17 @@ pub mod pallet {
 
 	#[pallet::pallet]
 	pub struct Pallet<T>(PhantomData<T>);
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			call_on_exposure!(Previous::iter_keys()
+				// TODO?: make this value adjustable
+				.drain()
+				.take(1)
+				.fold(Zero::zero(), |acc, e| acc + Self::payout_inner(e).unwrap_or(Zero::zero())))
+			.unwrap_or_default()
+		}
```

Comment:
```rs
#![allow(unused)]
fn main() {
    let mut v = vec![1, 2, 3, 4];
    let u: Vec<_> = v.drain(..).take(1).collect();
    assert_eq!(v, &[]);
    assert_eq!(u, &[1]);
}
```
The `drain` here will remove all items in the cache container. it's problematic.

Reply:
Rust `drain` is different from Substrate `drain`.

But all iterators are lazy.

You can try these:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b088f2f4942e6982803d607d07f453b

```rs
#[test]
fn playground() {
	ExtBuilder::default().collator_count(2).genesis_collator().build().execute_with(|| {
		dbg!(call_on_exposure_test!(Next::iter_keys().collect::<Vec<_>>()));
		call_on_exposure_test!(Next::iter_keys().drain().take(1).collect::<Vec<_>>());
		dbg!(call_on_exposure_test!(Next::iter_keys().collect::<Vec<_>>()));
	});
}
```